File: apps/web/src/api/admissibility.ts
import { api } from "./client";
import type { AdmissibilityAssessment, AdmissibilityContent } from "@/types";

/** Fetch the project's admissibility matrix (and generation status). */
export async function getAdmissibilityApi(projectId: string): Promise<AdmissibilityAssessment> {
  const { data } = await api.get(`/admissibility/project/${projectId}`);
  return data;
}

/** Queue AI generation of the admissibility matrix; returns immediately (poll GET). */
export async function generateAdmissibilityApi(projectId: string): Promise<{ status: string }> {
  const { data } = await api.post(`/admissibility/project/${projectId}/generate`);
  return data;
}

/** Persist an analyst-edited admissibility matrix. */
export async function saveAdmissibilityApi(
  projectId: string,
  content: AdmissibilityContent,
): Promise<AdmissibilityAssessment> {
  const { data } = await api.put(`/admissibility/project/${projectId}`, { content });
  return data;
}

File: apps/backend/app/api/v1/admissibility.py
import os

from fastapi import APIRouter, BackgroundTasks, Depends, HTTPException

from app.schemas.admissibility import AdmissibilitySave
from app.services import admissibility_service, project_service
from app.api.v1.deps import get_current_user

router = APIRouter()

_NOT_CONFIGURED = (
    "AI analysis is not configured — set ANTHROPIC_API_KEY in apps/backend/.env "
    "and restart the backend."
)


@router.get("/project/{project_id}")
async def get_assessment(project_id: str, _=Depends(get_current_user)):
    """The project's admissibility scoring matrix + generation status."""
    return admissibility_service.get(project_id)


@router.post("/project/{project_id}/generate")
async def generate_assessment(
    project_id: str, background: BackgroundTasks, current_user=Depends(get_current_user),
):
    """Queue AI generation of the project's admissibility matrix (background) and
    return immediately. The client polls GET /project/{id} for progress + result."""
    if current_user.get("role") == "Client View":
        raise HTTPException(status_code=403, detail="Clients cannot generate the admissibility matrix.")
    if not os.getenv("ANTHROPIC_API_KEY"):
        raise HTTPException(status_code=503, detail=_NOT_CONFIGURED)
    if not project_service.get_project(project_id):
        raise HTTPException(status_code=404, detail="Project not found")

    if admissibility_service.get(project_id).get("status") == "running":
        return {"status": "running"}

    admissibility_service.mark_running(project_id)
    background.add_task(admissibility_service.run_generation, project_id)
    return {"status": "running"}


@router.put("/project/{project_id}")
async def save_assessment(
    project_id: str, body: AdmissibilitySave, current_user=Depends(get_current_user),
):
    """Save an analyst-edited admissibility matrix."""
    if current_user.get("role") == "Client View":
        raise HTTPException(status_code=403, detail="Clients cannot edit the admissibility matrix.")
    if not project_service.get_project(project_id):
        raise HTTPException(status_code=404, detail="Project not found")
    return admissibility_service.save_content(project_id, body.content.model_dump())
